src/metabase/models/card.clj

```diff
@@ -594,29 +670,32 @@ saved later when it is ready."
   the transaction yet. If you pass true here it is important to call the event after the cards are successfully
   created."
   ([card creator] (create-card! card creator false))
-  ([{:keys [dataset_query result_metadata dataset parameters parameter_mappings], :as card-data} creator delay-event?]
+  ([{:keys [dataset_query result_metadata dataset parameters parameter_mappings type] :as card-data} creator delay-event?]
    ;; `zipmap` instead of `select-keys` because we want to get `nil` values for keys that aren't present. Required by
    ;; `api/maybe-reconcile-collection-position!`
    (let [data-keys            [:dataset_query :description :display :name :visualization_settings
-                               :parameters :parameter_mappings :collection_id :collection_position :cache_ttl]
-         card-data            (assoc (zipmap data-keys (map card-data data-keys))
-                                     :creator_id (:id creator)
-                                     :dataset (boolean (:dataset card-data))
-                                     :parameters (or parameters [])
-                                     :parameter_mappings (or parameter_mappings []))
+                               :parameters :parameter_mappings :collection_id :collection_position :cache_ttl :type]
+         card-data            (-> (zipmap data-keys (map card-data data-keys))
+                                  (assoc
+                                   :dataset    (or (when (= "model"type)
```

Comment:
This is a typo and a bug, you meant to write:
```suggestion
                                   :dataset    (or (when (= "model" type)
```
(aside: this should have been caught by tests)

If you make this change, it's still too easy to use `create-card!` incorrectly, in the same way as I pointed out with the API endpoint before. Here's a passing test that demonstrates what I consider risky behaviour:

```
(is (=?  {:type "model" :dataset true}
         (metabase.models.card/create-card! (merge (card-with-name-and-query (mt/random-name))
                                                   {:type "model" :dataset false})
                                            {:id (mt/user->id :crowberto)})))
```

It looks like `dataset` is being set to `false`, but under the covers it's being overridden to `true`. This should throw an error like the API endpoint. Since we need to calculate the value of `dataset?` after the update here, you can reuse `card/ensure-type-and-dataset-are-consistent` to do it so that the logic isn't duplicated.